pkg/controller/controlplane/valuesprovider.go

```diff
@@ -303,13 +333,79 @@ func isOverlayEnabled(networking *gardencorev1beta1.Networking) (bool, error) {
 }
 
 // getControlPlaneShootChartValues collects and returns the control plane shoot chart values.
-func (vp *valuesProvider) getControlPlaneShootChartValues(cluster *extensionscontroller.Cluster) (map[string]interface{}, error) {
+func (vp *valuesProvider) getControlPlaneShootChartValues(cluster *extensionscontroller.Cluster, cp *apismetal.ControlPlaneConfig) (map[string]interface{}, error) {
 	if cluster.Shoot == nil {
 		return nil, fmt.Errorf("cluster %s does not contain a shoot object", cluster.ObjectMeta.Name)
 	}
 
+	metallb, err := getMetallbChartValues(cp)
+	if err != nil {
+		return nil, err
+	}
 	return map[string]interface{}{
 		metal.CloudControllerManagerName: map[string]interface{}{"enabled": true},
+		metal.MetallbName:                metallb,
 	}, nil
+}
 
+// getMetallbChartValues collects and returns the CCM chart values.
+func getMetallbChartValues(
+	cpConfig *apismetal.ControlPlaneConfig,
+) (map[string]interface{}, error) {
+	if cpConfig.LoadBalancerConfig == nil || cpConfig.LoadBalancerConfig.MetallbConfig == nil {
+		return map[string]interface{}{
+			"enabled": false,
+		}, nil
+	}
+
+	for _, cidr := range cpConfig.LoadBalancerConfig.MetallbConfig.IPAddressPool {
+		if err := parseAddressPool(cidr); err != nil {
+			return nil, fmt.Errorf("invalid CIDR %q in pool: %w", cidr, err)
+		}
+	}
+
+	enableSpeaker := false
+	if cpConfig.LoadBalancerConfig.MetallbConfig.EnableSpeaker {
+		enableSpeaker = true
+	}
+	enableL2 := false
+	if cpConfig.LoadBalancerConfig.MetallbConfig.EnableL2Advertisement {
+		enableL2 = true
+	}
```

Comment:
Can't we assign `cpConfig.LoadBalancerConfig.MetallbConfig.EnableSpeaker` and `cpConfig.LoadBalancerConfig.MetallbConfig.EnableL2Advertisement` to our return object directly? I think the local variables and `if` statements can be removed.

Reply:
good point, I have fixed that